### PR TITLE
Issue 2710 wsapi ver lock

### PR DESCRIPTION
**Pull request by shane-tomlinson**
READY FOR FINAL REVIEW!

Large portions of @ozten's original code from f2adc7aad1c3321f398dd83e9349610ba8514897 as well as an updated version of @lloyd's PR #2284 to turn version() into an asynchronous function are used.

The asynchronous version() is used because generate_code_version was using the randomly generated SHA that is returned before git has finished fetching the code version from disk.

Instead of placing the SHA into the wsapi URL, it is placed into the BrowserID-Version HTTP header.

### CODE REVIEW

**Comment by ozten** [comment]
Looks good. I think having this be asynchonous instead of returning a random value is a huge improvement.

**Comment by lloyd** [comment]
let's merge #2284 first on monday, then rebase this.  preliminarily it looks good.

**Comment by jaredhirsch** [comment]
@lloyd @ozten @shane-tomlinson bump! anything holding this up?

**Comment by shane-tomlinson** [comment]
It is completely broken on ephemeral instances. New help. Description contains info.

Jared Hirsch notifications@github.com wrote:

@lloyd @ozten @shane-tomlinson bump! anything holding this up?

---

**Comment by lloyd** [comment]
only thing holding it up from my side is lack of time to review.  I'm eager to land the monster PR then hope to contribute to this one.

Aside from general review, the thing that is most important to me is to land this initially with zero behavioral change, just report a statsd counter when frontend and backend code doesn't match.

**Comment by callahad** [comment]
Oh man. I'm so sorry to do this. But as per [RFC 6648](http://tools.ietf.org/html/rfc6648), we should name the header `BrowserID-Version`, and _not_ use the `X-` prefix. This is especially important because we actually are, explicitly, aiming for future standardization.

File: lib/wsapi_client.js [hunk 1]
```diff
@@ -82,31 +91,36 @@ exports.get = function(cfg, path, context, getArgs, cb) {
...
   }
 
-  var headers = { };
-  injectCookies(context, headers);
-
-  if (typeof getArgs === 'object')
-    path += "?" + querystring.stringify(getArgs);
-
-  var timingHandle = startRequest();
-  meth.get({
```
Comment (shane-tomlinson) [inline lib/wsapi_client.js:92]: Removing the tweaker from the picture.

File: lib/wsapi_client.js [hunk 2]
```diff
@@ -82,31 +91,36 @@ exports.get = function(cfg, path, context, getArgs, cb) {
...
+      'X-BrowserID-Version': code_version
+    };
+    injectHeaders(context, headers);
+    injectCookies(context, headers);
+
+    if (typeof getArgs === 'object')
+      path += "?" + querystring.stringify(getArgs);
+
+    var timingHandle = startRequest();
+    meth.get({
```
Comment (shane-tomlinson) [inline lib/wsapi_client.js:105]: Only to put him back again.

File: lib/static/views.js [hunk 3]
```diff
@@ -100,6 +100,17 @@ exports.setup = function(app) {
     next();
   });
 
+  app.get('/build/code_version.js', function(req, res) {
```
Comment (ozten) [inline lib/static/views.js]: From a maintenance perspective, it is confusing that have both a route and a `scripts/generate_code_version.js` that generates a static file.

We should choose one or the other... no?

Perhaps this route could be a middleware and would call `scripts/generate_code_version.js` if there is no code_version.js on disc?

File: scripts/generate_code_version.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+const
+fs = require('fs'),
+path = require('path'),
+mkdirp = require('mkdirp'),
+logger = require('../lib/logging').logger,
+version = require('../lib/version.js');
+
+module.exports = function(cb) {
```
Comment (ozten) [inline scripts/generate_code_version.js]: Currently the signature of the callback is no parameters.

Change that to have `error` as the first parameter. There are conditions where this code can fail. We should bubble those failures up.

File: scripts/compress [hunk 5]
```diff
@@ -4,75 +4,79 @@ var
...
-computecluster = require('compute-cluster');
+computecluster = require('compute-cluster'),
+version = require('../lib/version.js');
 
 const staticPath = path.join(__dirname, '..', 'resources', 'static');
 
 var langs = config.get('supported_languages');
 
-var all = resources.all(langs);
+generateCodeVersionJS(function() {
```
Comment (ozten) [inline scripts/compress]: See note about adding err to callback function signature.

File: lib/static/views.js (hunk 3 above)
Comment (shane-tomlinson) [inline lib/static/views.js]: I did it that way initially and that felt yuck too. I'll have another look tomorrow. Thanks.

File: scripts/generate_code_version.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        if (err.code === "ENOENT") {
+          // file doesn't exist? no problem. No need to compare versions, just
+          // write it out.
+          return writeCodeVersion();
+        }
+
+        // Something more serious is happening, uh oh.
+        var msg = "Error reading " + codeVersionPath + ": " + String(err);
+        console.error(msg);
+        return logger.error(msg);
```
Comment (ozten) [inline scripts/generate_code_version.js]: Use callback and pass `msg` back.

File: tests/post-limiting-test.js [hunk 7]
```diff
@@ -22,29 +23,52 @@ suite.options.error = false;
 
 start_stop.addStartupBatches(suite);
 
+function addCodeVersionHeader(path, code_version, headers) {
+  // All wsapi calls require the BrowserID-Version header
+}
```
Comment (ozten) [inline tests/post-limiting-test.js]: Unused function? If so remove.

File: lib/wsapi.js [hunk 8]
```diff
@@ -138,6 +139,21 @@ function checkCSRF(req, resp, next) {
...
 
+function checkCodeVersion(req, resp, next) {
+  version(function(expectedCodeVersion) {
+    if (req.url.indexOf(WSAPI_PREFIX) === -1) return next();
+    var requestedCodeVersion = req.headers['browserid-version'];
+
+    if (requestedCodeVersion !== expectedCodeVersion) {
+      logger.warn("Code version mis-match: " + req.url + " expected: " + expectedCodeVersion + " received: " + requestedCodeVersion);
+
+      statsd.increment("wsapi_code_mismatch." + req.url);
```
Comment (ozten) [inline lib/wsapi.js:150]: Shouldn't we respond with a 400 or other "bad client request' HTTP status code?

File: tests/session-context-test.js [hunk 9]
```diff
@@ -99,6 +99,20 @@ suite.addBatch({
   }
 });
 
+suite.addBatch({
+  "code version mismatch using session_context API": {
+    topic: wsapi.get('/wsapi/session_context', {}, {
+      headers: {
+        "Browserid-Version": "INVALID_SHA"
```
Comment (ozten) [inline tests/session-context-test.js:106]: Why is session_context exempt from version lock?

**Comment by ozten** [comment]
Great work!

File: tests/session-context-test.js (hunk 9 above)
Comment (shane-tomlinson) [inline tests/session-context-test.js:106]: It's not. The code that I snagged from you returned a 400 response if the version was incorrect. @lloyd asked that the behavior be changed so if a version mismatch occurs, the wsapi still responds normally but a statsd log entry is added.

This test makes it easy to test for changes to this behavior. When we want to disallow version mismatches, change 200 to whatever the new expected response code is.

File: tests/post-limiting-test.js (hunk 7 above)
Comment (shane-tomlinson) [inline tests/post-limiting-test.js]: Good call. That is left over from a refactor.

File: scripts/generate_code_version.js (hunk 4 above)
Comment (shane-tomlinson) [inline scripts/generate_code_version.js]: Good call.

File: scripts/compress (hunk 5 above)
Comment (shane-tomlinson) [inline scripts/compress]: On it.

File: lib/wsapi.js (hunk 8 above)
Comment (shane-tomlinson) [inline lib/wsapi.js:150]: @lloyd asked for the behavior that you wrote to be changed until we have a better understanding of how frequently this is occurring. All he wants for now is to tick a statsd counter, but continue to process the wsapi normally.

File: lib/static/views.js (hunk 3 above)
Comment (shane-tomlinson) [inline lib/static/views.js]: I think I have a cleaner solution for this now.

**Comment by ozten** [comment]
Awesome! r+

**Pull request merged**